### ISSUE TITLE
Certificate error (-201)

### ISSUE THREAD

**Issue by last0tt0man** [RootMyTV/RootMyTV.github.io#9]
Hi ;

TV Model : 49SJ800V
Firmware : 06.00.25

I'm trying every combination ( Change Country , time and date , downgrade firmware to 05.80.35, etc ) but 
Unable to Load -201 certificate error everytime.

How to fix this issue ??

**Comment by NonStandardModel** [comment]
Same for me. 
Note: old webos verion (3.3.4) with fw 05.50.55 (offering update to 05.60.25).
It is interesting that already connecting to rootmy.tv will give Browser warning (here we can click "proceed anyway"), but later in download stage it gets blocked. When I click "try again" it will just go to black-screen and nothing happens.

**Comment by Informatic** [comment]
Seems like webOS was only trusting Let's Encrypt's DST Root CA X3 which
expired on 2021/09/30. I'll try to find some sensible solution soon.

**Comment by tekand** [comment]
Plex and Emby users reported the same problem with Let's Encrypt not trusted in below webos 5.0.
Some Emby users had success with using zerossl certs.

https://emby.media/community/index.php?/topic/102144-several-lg-tvs-cannot-connect-to-server/page/2/

Also does this need https to work?

**Comment by Informatic** [comment]
Yeah, I was going to set up a workaround for that by using some other CA.
I'll try to fix this by the end of the week.

https is required for entering stage2 - we could theorretically use some
open redirector trusted by older WebOS, but then we also don't feel
comfortable hosting exploits over unencrypted http.

**Comment by tekand** [comment]
Understood. Thanks for your efforts. I came here because my Plex client broke on the LG TV and was trying to see how can I update the root certs on it.

**Comment by Informatic** [comment]
Hey, could you test out this domain? This just proxies traffic over to main rootmy.tv website for now, but presents itself with zerossl/usertrust-issued certificate.

https://rootmytv.inf.re

**Comment by tekand** [comment]
Maybe this needs to be updated as well: ?
https://github.com/RootMyTV/RootMyTV.github.io/blob/364e67ce5a02636140531995775b14efc26c2ad3/index.html#L55

**Comment by Informatic** [comment]
This URL is only used as a fallback when rootmytv exploit is downloaded and launched from local filesystem. In such case one may want to replace it locally anyway.

Also we're most probably going to point https://rootmy.tv domain at that server in the end.

**Comment by cameldebrno** [comment]
It's working now 👍

**Comment by tekand** [comment]
```
start-devmode.sh: Downloading from https://rootmytv.inf.re/files/stage3.sh...
err: {"retrunValue":false, "errorCode": -1, "errorText": "Denied method call \"download\" for category \"/\""}
```
:(

**Comment by The-Dude-2020** [comment]
Homebrew channel installed, but now it sais: An error occured while downloading repository.

**Comment by bjornreppen** [comment]
> Hey, could you test out this domain? This just proxies traffic over to main rootmy.tv website for now, but presents itself with zerossl/usertrust-issued certificate.
> 
> https://rootmytv.inf.re

This URL still tries to load stage3.sh from https://rootmy.tv (not .inf.re) and fails with peer certificate cannot be authenticated.

**Comment by Informatic** [comment]
That's weird. stage3 URL is built off stage2 domain already (and stage2 *wouldn't* load off a domain with untrusted certificate - as per OPs issue) - https://github.com/RootMyTV/RootMyTV.github.io/blob/main/stage2.html#L32

I confirmed it's working on webOS 3.4 and webOS 3.8 here... (where https://rootmy.tv fails)

**Comment by tekand** [comment]
It works on my webOS 4.4 as well.

**Comment by bjornreppen** [comment]
I experienced a little weird behaviour. When opening https://rootmytv.inf.re and sliding right I get the certificate fail message mentioning rootmy.tv host, but when I proceed to press the remote control back button it seems to do what it is supposed to.  On my 55NANO95 it installed successfully after back press, but on my OLED65CX it states download complete, but target directory did not match. TV is likely not vulnerable.  Not sure why pressing the BACK button would help, but it seems to do something on both my TVs

**Comment by NonStandardModel** [comment]
Works via https://rootmytv.inf.re for me (rootmy.tv still fails). Homebrew installed OK, not its homebrew repo that fails to download.

**Comment by The-Dude-2020** [comment]
> Works via https://rootmytv.inf.re for me (rootmy.tv still fails). Homebrew installed OK, not its homebrew repo that fails to download.

The same thing happens to me.

**Comment by The-Dude-2020** [comment]
> Also - regarding early boot commands - it's preferable to add extra executables to `/var/lib/webosbrew/init.d` - all executable scripts in this directory are run on startup (also they are guarded by "failsafe mode" - the TV will _not_ run these, in case something goes wrong and a reboot/crash happens when executing them)
> 
> Haven't tested this, but this should work:
> 
> ```shell
> mkdir -p /var/lib/webosbrew/init.d /media/cryptofs/root
> curl -k https://letsencrypt.org/certs/isrgrootx1.pem.txt | tee /media/cryptofs/root/isrgrootx1.pem.txt
> echo "mount -o bind /media/cryptofs/root/isrgrootx1.pem.txt /usr/share/ca-certificates/mozilla/DST_Root_CA_X3.crt" > /var/lib/webosbrew/init.d/50-letsencrypt-hack
> chmod +x /var/lib/webosbrew/init.d/50-letsencrypt-hack
> ```

So basically add these to init.d and reboot the tv?

**Comment by Informatic** [comment]
Just execute these commands manually once. These will:
1. create `/var/lib/webosbrew/init.d` and `/media/cryptofs/root` directories
2. download ISRG Root X1 cert to `/media/cryptofs/root/isrgrootx1.pem.txt` file
3. create a script in `/var/lib/webosbrew/init.d/50-letsencrypt-hack` that'll mount over the certificate into an apropriate directory
4. make that script executable so it is run on every boot

After that you'll need to reboot the TV and it *should* work fine.

**Comment by The-Dude-2020** [comment]
> Just execute these commands manually once. These will:
> 
> 1. create `/var/lib/webosbrew/init.d` and `/media/cryptofs/root` directories
> 2. download ISRG Root X1 (the "new" Let's Encrypt one) cert to `/media/cryptofs/root/isrgrootx1.pem.txt` file
> 3. create a script in `/var/lib/webosbrew/init.d/50-letsencrypt-hack` that'll mount over the certificate into an apropriate directory
> 4. make that script executable so it is run on every boot
> 
> After that you'll need to reboot the TV and it _should_ work fine.

Cool, thanks for your help.

**Comment by The-Dude-2020** [comment]
It works like a charm 👍🥳

**Comment by pannal** [comment]
Same issues with my LG B9 on 05.10.30 (main domain results in certificate error, second one fails on stage 3)

**Comment by kesenwang** [comment]
I'm getting the "This likely means your TV is not vulnerable to LunaDownloadMgr exploit" on the suggested domain, but on Rootmy.tv I get the same certificate error shown above.

**Comment by Informatic** [comment]
We're rolling out fixed certificates https://rootmy.tv - should be working in a couple of hours.
Any issues now (`[...]not vulnerable[...]`/`Denied method call` errors in stage3) are caused by patched firmware versions, see #2.
